jax/interpreters/pxla.py

```diff
@@ -47,14 +48,15 @@
 
 def identity(x): return x
 
-def shard_args(backend, devices, assignments, axis_size, tuple_args, args):
+def shard_args(backend, devices, indices, axis_size, tuple_args, args):
   """Shard each argument data array along its leading axis.
 
   Args:
     backend: the platform to be used
     devices: list of Devices mapping replica index to a physical device.
-    assignments: list of integers with the same length as `devices` mapping
-      replica index to an index along the leading axis (i.e. a shard).
+    indices: list of logical indices into the argument with the same length as
+      `devices`. Can be an int, a slice object, or a tuple thereof. This
```

Comment:
How about tweaking the last two sentences to: "Each element of the list can be an int, a slice, or a tuple thereof. This describes how logical input arrays should be sharded/replicated to physical device buffers: each list element indicates the portion of the array is represented on the corresponding device."

That leaves me with some questions though, around the constraints `indices` must satisfy to be valid.

Previously, when `assignments` was a list of integers, the integer values had to be collectively exhaustive in the sense that every row of the input array was covered, i.e. `set(assignments) == set(range(axis_size))`. Is that constraint still true now?

Also, previously the amount of replication was a constant, i.e. `len(set(Counter(assignments).values())) == 1`. Is there any corresponding constraint about replication now?

Finally, we had (and I think still have) the constraint that the leading axis size of each argument is `axis_size`, which is why we only needed one `assignments` that could apply to each element of `args`. That is, since `all(x.shape[0] == axis_size for x in args if x is not core.unit)`, we knew that having `set(assignments) == set(range(axis_size))` was enough to ensure `assignments` described a sharding for all input args (only along their leading axes). What constraints exist between the elements of `args` now? For example, when an element of `indices` is a tuple, does that mean all args have to have rank (number of axes) greater than or equal to the length of that tuple?

I'm not suggesting the docstring needs to have answers to all these questions; I'm asking them about the design. But we probably should have answers written down somewhere (e.g. in comments, design docs, pr messages). Maybe the `ShardedDeviceArray` docstring?

Reply:
These are all great points!

> How about tweaking the last two sentences to:

I added an abridged version of your suggestion and refer back to ShardedDeviceArray.logical_indices.

 > the integer values had to be collectively exhaustive

This is still true, in that the indices must cover the full logical array. I added this to the ShardedDeviceArray.logical_indices description. It would also be good to assert this in the SDA constructor, although I wanna add a benchmark before adding too much stuff to SDA creation. I added a TODO for now.

> What constraints exist between the elements of args now? For example, when an element of indices is a tuple, does that mean all args have to have rank (number of axes) greater than or equal to the length of that tuple?

Good question :) Each arg's rank does have to be >= the number of subindices (that's a word I made up earlier today for the elements inside an index tuple). This is gonna become more complicated with replication and partitioning though. Unfortunately, I think we're gonna need different indices for different arguments, since arguments can be individually replicated or partitioned (or pmapped once James has his way). The only constraint will be that each argument is ultimately distributed onto the same number of devices.

> we probably should have answers written down somewhere

Agreed. I'd like the code to be as self-documenting as possible, although a design doc or two describing the overall problem and possible solutions wouldn't hurt either. The details/invariants are also gonna change as we iterate and add features (e.g. maybe we'll canonicalize all indices into full-rank tuples, like James suggested). Please keep asking these questions!